torch/fx/experimental/gradual_graph_typechecker.py

```diff
@@ -0,0 +1,90 @@
+from functools import reduce
+from torch.fx.tensor_type import Dyn, consistency, Tensor_Type
+
+
+def type_check(graph):
+    """
+    A gradual type checker for graphs
+    """
+    env = {}
+    expr_env = {}
+    # populate the type environment
+    collect_params(graph.nodes, env)
+    for n in graph.nodes:
+        if not type_check_node(n, env, expr_env):
+            return False
+    return True
+
+
+def collect_params(node_list, env):
+    """
+    Collect argument types from the type environment.
+    If not present, populate with Dyn
+    """
+    for node in node_list:
+        if node.op == 'placeholder':
+            if node.type:
+                env[node.name] = node.type
+            else:
+                node.type = Dyn
+                env[node.name] = node.type
+
+
+def type_check_node(n, env, expr_env):
+    """
+    Type check a given node.
+    Current operations:
+    - Reshape
+    - Transpose
+    - Add
+    """
+    if n.op == 'placeholder':
+        if n.name in env.keys():
+            n.type = env[n.name]
+            return n.type
+        else:
+            raise Exception("free variable")
+
+    if n.op == 'call_function':
+        if n.name == 'add':
+            t1 = type_check_node(n.args[0], env, expr_env)
+            t2 = type_check_node(n.args[1], env, expr_env)
+            if consistency(t1, t2):
+                expr_env[n.name] = t1
+                n.type = t1
+                return n.type
+            else:
+                return False
+
+        if n.name == 'reshape':
```

Comment:
Names are not load-bearing, they're only used for codegen and debugging. What you want to check is the target, e.g. `if n.target == torch.reshape`